src/tracing/esp32_diagnostic_trace/Diagnostics.h

```diff
@@ -0,0 +1,127 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/core/TLVCircularBuffer.h>
+
+namespace chip {
+namespace Tracing {
+
+namespace Diagnostics {
+
+/**
+ * @class DiagnosticEntry
+ * @brief Abstract base class for encoding diagnostic entries into TLV format.
+ *
+ */
+class DiagnosticEntry
+{
+public:
+    /**
+     * @brief Virtual destructor for proper cleanup in derived classes.
+     */
+    virtual ~DiagnosticEntry() = default;
+
+    /**
+     * @brief Pure virtual method to encode diagnostic data into a TLV structure.
+     *
+     * @param writer A reference to the `chip::TLV::CircularTLVWriter` instance
+     *               used to encode the TLV data.
+     * @return CHIP_ERROR Returns an error code indicating the success or
+     *                    failure of the encoding operation.
+     */
+    virtual CHIP_ERROR Encode(chip::TLV::CircularTLVWriter & writer) = 0;
+};
+
+template <typename T>
+class Diagnostic : public DiagnosticEntry
+{
+public:
+    Diagnostic(const char * label, T value, uint32_t timestamp) : label_(label), value_(value), timestamp_(timestamp) {}
+
+    CHIP_ERROR Encode(chip::TLV::CircularTLVWriter & writer) override
+    {
+        chip::TLV::TLVType DiagnosticOuterContainer = chip::TLV::kTLVType_NotSpecified;
+        ReturnErrorOnFailure(
+            writer.StartContainer(chip::TLV::AnonymousTag(), chip::TLV::kTLVType_Structure, DiagnosticOuterContainer));
+        ReturnErrorOnFailure(writer.Put(chip::TLV::ContextTag(0), timestamp_));
```

Comment:
we should define these tag fields with some constants, it would be easier to understand and the same enums can be used in the python script that is utilizing this data.